### DISCUSSION TITLE
ENH: Manage "Indicator Constraints"

### DISCUSSION THREAD

**Discussion by mingodad** [ERGO-Code/HiGHS#859] [Ideas]
Testing highs with miplib2017 models (right now only LP) there is several models that highs can not handle due to the "INDICATORS" section present on the mps file.
Looking here https://www.gams.com/latest/docs/UG_LanguageFeatures.html#UG_LanguageFeatures_IndicatorConstraints for the explanation of it.
Would be nice if HiGHS could implement it !

**Comment by jajhall** [comment]
Sorry to take so long to respond, but please could you let us know instance names with an INDICATORS section

**Comment by mingodad** [comment]
Using this script in the root of `miplib2017` from http://miplib.zib.de I've got the trimmed down output bellow:
```
#!/bin/sh
for fn in mps/*
do
    echo $fn
    zcat $fn | grep INDICATORS
done
```
Output:
```
mps/amaze22012-03-15i.mps.gz
INDICATORS
mps/amaze22012-06-28i.mps.gz
INDICATORS
mps/amaze22012-07-04i.mps.gz
INDICATORS
mps/cvrpa-n64k9vrpi.mps.gz
INDICATORS
mps/cvrpb-n45k5vrpi.mps.gz
INDICATORS
mps/cvrpp-n16k8vrpi.mps.gz
INDICATORS
mps/cvrpsimple2i.mps.gz
INDICATORS
mps/diameterc-mstc-v20a190d5i.mps.gz
INDICATORS
mps/diameterc-msts-v40a100d5i.mps.gz
INDICATORS
mps/elitserienhandball11i.mps.gz
INDICATORS
mps/elitserienhandball13i.mps.gz
INDICATORS
mps/elitserienhandball14i.mps.gz
INDICATORS
mps/elitserienhandball3i.mps.gz
INDICATORS
mps/fillomino7x7-0i.mps.gz
INDICATORS
mps/fjspeasy01i.mps.gz
INDICATORS
mps/gfd-schedulen180f7d50m30k18-16i.mps.gz
INDICATORS
mps/gfd-schedulen180f7d50m30k18.mps.gz
mps/gfd-schedulen25f5d20m10k3i.mps.gz
INDICATORS
mps/gfd-schedulen55f2d50m30k3i.mps.gz
INDICATORS
mps/ghoulomb4-9-10i.mps.gz
INDICATORS
mps/l2p1i.mps.gz
INDICATORS
mps/l2p2i.mps.gz
INDICATORS
mps/mappingmesh3x3mpeg2i.mps.gz
INDICATORS
mps/mario-t-hard5i.mps.gz
INDICATORS
mps/mrcpspj30-15-5i.mps.gz
INDICATORS
mps/mrcpspj30-17-10i.mps.gz
INDICATORS
mps/mrcpspj30-53-3i.mps.gz
INDICATORS
mps/msc98-ip.mps.gz
mps/mspsphard01i.mps.gz
INDICATORS
mps/mspsphard03i.mps.gz
INDICATORS
mps/oocsp-racks030e6cci.mps.gz
INDICATORS
mps/oocsp-racks030f7cci.mps.gz
INDICATORS
mps/pizza27i.mps.gz
INDICATORS
mps/pizza78i.mps.gz
INDICATORS
mps/rpp22falsei.mps.gz
INDICATORS
mps/shipschedule3shipsi.mps.gz
INDICATORS
mps/shipschedule6shipsmixi.mps.gz
INDICATORS
mps/shipschedule8shipsmixuci.mps.gz
INDICATORS
mps/splice1k1i.mps.gz
INDICATORS
mps/stoch-vrpvrp-s5v2c8vrp-v2c8i.mps.gz
INDICATORS
mps/supportcase21i.mps.gz
INDICATORS
mps/supportcase27i.mps.gz
INDICATORS
mps/supportcase28i.mps.gz
INDICATORS
```

**Comment by jajhall** [comment]
Thanks :-)

**Comment by jajhall** [comment]
A user sent me a .lp file with indicator constraints recently because HiGHS failed silently when reading it. This got me thinking about how to implement them in general, as I had to guide her on how to model them explicitly so that HiGHS could solve her problem. Although they'd still have to be handled via "big-M" reformulation - rather than by handling them explicitly in the MIP solver - I'm inclined to add them to HiGHS

**Comment by jajhall** [comment]
Inspired by Erling Andersen's discussion of handling various MIP extensions as special cases of disjunctive constraints, implementing this in HiGHS is closer to becoming a reality.

**Reply by jhmenke** [reply]
Are indicator constraints currently implemented in HiGHS?

**Reply by jajhall** [reply]
Not yet

**Reply by odow** [reply]
@jhmenke if you're using JuMP and your `x` variables have finite bounds: you can use indicator constraints `@constraint(model, z --> {a' * x <= b})`
